### PR TITLE
Reindex rethrottle persistent task

### PR DESCRIPTION
**Pull request by henningandersen**
This adds support for rethrottling resilient/persistent reindex
through updating the .reindex index and notifying the task. This
ensures that the new throttle value sticks on failovers while
also ensuring that the task wakes up immediately if it had a very
low throttle value.

Related to #42612

This is the replacement for #50854.

Tests on the transport layer are missing, but I wanted to ensure
we agree on approach first and then tackle adding those
in a follow-up.

### CODE REVIEW

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTaskStateUpdater.java [hunk 1]
```diff
@@ -43,31 +45,36 @@
 
     private final ReindexIndexClient reindexIndexClient;
     private final ThreadPool threadPool;
-    private final String taskId;
+    private final String persistentTaskId;
     private final long allocationId;
+    private final TaskId ephemeralTaskId;
     private final ActionListener<ReindexTaskStateDoc> finishedListener;
     private final Runnable onCheckpointAssignmentConflict;
+    private Consumer<Float> requestsPerSecondObserver;
```
Comment (tbrooks8) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTaskStateUpdater.java]: Do we need this? I feel like adding the "poll" interaction model when we are doing all the working to implement the "push" component makes the interaction with the state updater and task more complicated that required.

**Review by Tim-Brooks — COMMENTED** [review]
I think this has a bug currently. 

`BulkByScrollParallelizationHelper#sendSubRequests` will use the requests per second parameter in `request.forSlice` which means it will use the original value (not updated) on a failover.

I think you are attempting to avoid mutating the request inside the index (and storing an additional request per second field) to retain the original request? Is that of value? We could store like a rethrottle audit field and mutate the request. So there is like a list field showing the submitted throttle values (and maybe time) and update the request to whatever the current value is.

**Comment by tbrooks8** [comment]
You could also mutate the request immediately when putting it out of the index and still have a dedicated field for the requests per second.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTaskStateUpdater.java (hunk 1 above)
Comment (henningandersen) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTaskStateUpdater.java]: Thanks, I agree, removed in b6096a9398a458273c392fcb832e140eaaa73fb2

**Comment by henningandersen** [comment]
> I think this has a bug currently.

That was kind of on purpose, since slicing is not supported for resilient reindex anyway.

I changed it to modify the request instead, that is a bit nicer.

**Review by Tim-Brooks — COMMENTED** [review]
This mostly looks good to me. I don't see anything that needs to change. Let me just take one last look tomorrow when I have a bit more time. Also, I think it needs to be updated now that I have merged the persistent task stuff.

**Comment by henningandersen** [comment]
Test failure reported here #51926, so:
@elasticmachine run elasticsearch-ci/1

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Pull request merged**